### ISSUE TITLE
Add basic auth support for proxies in the plugin manager

### ISSUE THREAD

**Issue by clintongormley** [elastic/elasticsearch#17046]
The plugin manager script supports proxies with the `-Dhttp.proxyHost` and `-Dhttp.proxyPort` settings. Unfortunately this does not support authenticated proxies out of the box, but it looks fairly easy for us to add this by setting the default authenticator in eg the plugin cli main when we see `-Dhttp.proxyUser` and `-Dhttp.proxyPassword`.

From https://github.com/elastic/elasticsearch/issues/16633#issuecomment-183489149

**Comment by sarwarbhuiyan** [comment]
Does this work too?

URL url = new URL(“location address”);
URLConnection uc = url.openConnection();
String userpass = username + ":" + password;
String basicAuth = "Basic " + new String(new Base64().encode(userpass.getBytes()));
uc.setRequestProperty ("Authorization", basicAuth);
InputStream in = uc.getInputStream();

Just considering whether that might be better than calling Authenticator.setDefault. Thoughts?

**Comment by rjernst** [comment]
@sarwarbhuiyan That works, except that we still have to read the sysprops. I guess I think it is cleaner to have this behave as if java already knew about these sysprops, and do their reading/handling outside of the code which is opening the url (so it has no knowledge of proxies at all).

**Comment by rjernst** [comment]
Note too that we have 2 places in the install plugin command that get a URL, the plugin zip itself, but also the sha1 hash for official and maven plugins.

**Comment by sarwarbhuiyan** [comment]
Ok, I'm with you. setting the static authentication doesn't require
touching the existing code or any new code that may forget to set it.
What's the best place to initialise this?

**Comment by rjernst** [comment]
IMO, inside the main method for plugins cli.  But I am cautious about adding it until we can actually test it, which mean we need a fixture that can mock the proxy. I worked on a branch with @rmuir (https://github.com/rjernst/elasticsearch/tree/separate_plugin_cli) that has an https server for integ tests. We could do something similar in another after that is in (or maybe even in the same tests, since we can have a parallel fixture that acts as the proxy (but doesn't need to be a real proxy, we only need to know it hit the endpoint and used the user/pass auth we specified).
